tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/AspNetCoreBlockMiddlewareIntegrationEnd.cs

```diff
@@ -0,0 +1,73 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#if !NETFRAMEWORK
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.DependencyInjection;
+#endif
+using System.Diagnostics;
+using Datadog.Trace.AppSec;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AspNetCore
+{
+    /// <summary>
+    /// The ASP.NET Core middleware integration.
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = AssemblyName,
+        TypeName = ApplicationBuilder,
+        MethodName = "Build",
+        ReturnTypeName = ClrNames.Void,
+        MinimumVersion = Major3,
+        MaximumVersion = Major6,
+        IntegrationName = nameof(IntegrationId.AspNetCore))]
+    [InstrumentMethod(
+        AssemblyName = AssemblyName,
+        TypeName = InternalApplicationBuilder,
+        MethodName = "Build",
+        ReturnTypeName = "Microsoft.AspNetCore.Http.RequestDelegate",
+        MinimumVersion = Major2,
+        MaximumVersion = Major2,
+        IntegrationName = nameof(IntegrationId.AspNetCore))]
+    public static class AspNetCoreBlockMiddlewareIntegrationEnd
+    {
+        private const string Major2 = "2";
+        private const string Major3 = "3";
+        private const string Major6 = "6";
+        private const string AssemblyName = "Microsoft.AspNetCore.Http";
+
+        private const string ApplicationBuilder = "Microsoft.AspNetCore.Builder.ApplicationBuilder";
+        private const string InternalApplicationBuilder = "Microsoft.AspNetCore.Builder.Internal.ApplicationBuilder";
+
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <returns>Calltarget state value</returns>
+        public static CallTargetState OnMethodBegin<TTarget>(TTarget instance)
+        {
+#if !NETFRAMEWORK
+            if (Security.Instance.Settings.Enabled)
+            {
+                var appb = (IApplicationBuilder)instance;
+                // make sure this is the very last call to build() as build can be called *many* times before the last one.
+                // theoretically, we should have 1. onmethodbegin, 2. CallTarget.Handlers.Beginmethodhandler, 3. CalltargetInvoker.BeginMethod 4.ApplicationBuilder.Build, 5.GenericWebHostService.StartAsync, 6.AsyncMethodBuilderCore.Start<GenericWebHostService>, when comes into play, that's last call to ApplicationBuilder.
+                // todo: make more robust
+                var frame = new StackTrace().GetFrame(6);
```

Comment:
Yikes, this is _very_ expensive in the startup loop. I really don't think we want to do this 🙁 Also, we can't guarantee this callstack - every time _anyone_ calls `Map`, `ApplicationBuilder.Build()` is invoked. which could be an arbitrary stack depth.

What are you trying to do here? Ensure that this is the _first_ middleware in the pipeline? Or ensure it's added to every branch in the pipeline?

If it's the former, I think there are a couple of approaches you could take:

- Instrument `GenericWebHostService.StartAsync`. Add an additional entry to the start of `StartupFilters` which is a reverse duck typing on `IStartupFilter`. This will ensure you're the first filter in the pipeline.
- Alternatively, Instrument `HostingApplication` and "wrap" the RequestDelegate with your middleware. I think Zach did some work with RequestDelegates in a branch, so you could ask him about this approach.

Also, bear in mind that in .NET Core 2.1, the `GenericWebHostService` might _not_ be used. it might be using the legacy `WebHost` instead (we don't necessarily have current tests for this)

Reply:
>Also, we can't guarantee this callstack - every time anyone calls Map, ApplicationBuilder.Build() is invoked. which could be an arbitrary stack depth.

We actually want to ensure it's the _last_ middleware in the pipeline actually (`AspNetCoreBlockMiddlewareIntegration` is here to ensure it's the first too), so by checking the class `AsyncMethodBuilderCore` is called (other times it's not the same path), then we know it's the last call to the `Build` method.. but I agree it's not ideal..
I'm not exactly sure what to instrument to make sure it's the last MapWhen middleware inserted...  👀 
we need to get ahold of the `IApplicationBuilder` after the `Startup.Configure` has been called.. and it needs to be a `MapWhen` middleware type, otherwise it doesn't diverge from the main pipeline...

Reply:
> We actually want to ensure it's the last middleware in the pipeline actually 

Hmmm, that seems strange to me, why? 🤔 

If it's the last middleware in the pipeline, it'll _only_ be called if no other middleware returns a response, which means it will _only_ be called when it's a 404 response, in which case the block middleware doesn't seem to add any value.

I would have thought that you would want/need to do one of the following:

- Add the blocking branch between _every_ middleware. This is similar to the way the MiddlewareAnalysisMiddleware works, and seems like it would be the most robust, but also the most expensive approach ([I've written some blog posts on how this works in case it gives any ideas](https://andrewlock.net/tag/middleware-analysis-package/))
- Add the blocking branch just before the call to `UseEndpoints` (for .NET Core 3+) or UseMvc() (for .NET Core 2x). This is _typically_ where the endpoint logic is executed, so presumably is the last place you need to do the blocking?